File: prolog.c
/* Adapted from https://www.cl.cam.ac.uk/~am21/research/funnel/prolog.c */
/* prolog.c: a simple Prolog interpreter written in C++,                */
/*           including an example test run as main().                   */
/* Copyright (c) Alan Mycroft, University of Cambridge, 2000.           */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Misc helpers */

void *prover_malloc(size_t n) {
  void *ptr = malloc(n);
  if (!ptr)
    puts("\nOut of Memory.\n");
  return ptr;
}

void indent(size_t indents, const size_t size) {
  for (size_t i = 0; i < indents; i++)
    for (size_t j = 0; j < size; j++)
      putchar(' ');
}

typedef struct {
  size_t t1;
  size_t t2;
} Timestamp;
static Timestamp timestamp = {.t1 = 0, .t2 = 0};
static inline Timestamp timestamp_next() {
  if ((timestamp.t1 += 1) == 0)
    timestamp.t2 += 1;
  return timestamp;
}

static inline bool timestamp_cmp(Timestamp ts1, Timestamp ts2) {
  if (ts1.t2 > ts2.t2)
    return true;
  if (ts1.t1 > ts2.t1)
    return true;
  return false;
}

/****************/
/* Declarations */
/****************/

typedef char *Atom;
union Term;
typedef union Term Term;
struct Goal;
typedef struct Goal Goal;
struct Clause;
typedef struct Clause Clause;
struct Program;
typedef struct Program Program;

/********/
/* Term */
/********/

typedef struct {
  size_t arity;
  Atom fsym;
  Term *subterms;
} TermConstant;

typedef struct {
  Term *instance;
  Timestamp timestamp;
} TermVariable;

typedef union Term {
  union {
    TermVariable tv; // tc_flag = 0
    TermConstant tc; // tc_flag = 1
  };
  bool tc_flag;
} Term;

/* Claims ownership of subterms */
Term *TermCons_init(Term *self, Atom f, Term *subterms, size_t arity) {
  self->tc.fsym = f;
  self->tc.subterms = subterms;
  self->tc.arity = arity;
  self->tc_flag = true;
  return self;
}

Term *TermVar_init(Term *self) {
  self->tc_flag = false;
  self->tv.instance = self;
  self->tv.timestamp = timestamp_next();
  return self;
}

void Term_print(Term *self) {}
bool Term_unify(Term *self, Term *other) {}
Term *Term_copy(Term *self) {
  if (self->tc_flag) {
    size_t arity = self->tc.arity;
    char *buf = prover_malloc(sizeof(Term) + sizeof(Term) * arity);
    Term *cpy = (Term *)buf;
    Term *subterms = cpy + 1;
    memcpy(subterms, self->tc.subterms, self->tc.arity);
    return TermCons_init(cpy, self->tc.fsym, subterms, arity);
  } else {
    return TermVar_init((Term *)prover_malloc(sizeof(Term)));
  }
}

/********/
/* Goal */
/********/

/**********/
/* Clause */
/**********/

/***********/
/* Program */
/***********/

class Term {
public:
  virtual void print() = 0;

public:
  virtual bool unify(Term *) = 0;

public:
  virtual bool unify2(TermCons *) = 0;

public:
  virtual Term *copy() = 0;
};

class TermCons : public Term {
private:
  int arity;
  Atom *fsym;
  Term **args;

public:
  void print() {
    fsym->print();
    if (arity > 0) {
      cout << "(";
      for (int i = 0; i < arity;) {
        args[i]->print();
        if (++i < arity)
          cout << ",";
      }
      cout << ")";
    }
  }
  bool unify(Term *t) { return t->unify2(this); }
  Term *copy() { return copy2(); }
  TermCons *copy2() { return new TermCons(this); }

private:
  TermCons(TermCons *p)
      : fsym(p->fsym), arity(p->arity),
        args(p->arity == 0 ? NULL : new Term *[p->arity]) {
    for (int i = 0; i < arity; i++)
      args[i] = p->args[i]->copy();
  }
  bool unify2(TermCons *t) {
    if (!(fsym->eqatom(t->fsym) && arity == t->arity))
      return false;
    for (int i = 0; i < arity; i++)
      if (!args[i]->unify(t->args[i]))
        return false;
    return true;
  }
};

class TermVar : public Term {
private:
  Term *instance;
  int varno;
  static int timestamp;

public:
  TermVar() : instance(this), varno(++timestamp) {}
  void print() {
    if (instance != this)
      instance->print();
    else
      cout << "_" << varno;
  };
  bool unify(Term *t);
  Term *copy();
  Term *reset() { instance = this; }

private:
  bool unify2(TermCons *t) { return this->unify(t); }
};

class Program;
class TermVarMapping;

class Goal {
private:
  TermCons *car;
  Goal *cdr;

public:
  Goal(TermCons *h, Goal *t) : car(h), cdr(t) {}
  Goal *copy() {
    return new Goal(car->copy2(), cdr == NULL ? NULL : cdr->copy());
  }
  Goal *append(Goal *l) {
    return new Goal(car, cdr == NULL ? NULL : cdr->append(l));
  }
  void print() {
    car->print();
    if (cdr != NULL) {
      cout << "; ", cdr->print();
    }
  }
  void solve(Program *p, int level, TermVarMapping *map);
};

class Clause {
public:
  TermCons *head;
  Goal *body;
  Clause(TermCons *h, Goal *t) : head(h), body(t) {}
  Clause *copy() {
    return new Clause(head->copy2(), body == NULL ? NULL : body->copy());
  }
  void print() {
    head->print();
    cout << " :- ";
    if (body == NULL)
      cout << "true";
    else
      body->print();
  }
};

class Program {
public:
  Clause *pcar;
  Program *pcdr;
  Program(Clause *h, Program *t) : pcar(h), pcdr(t) {}
};

class Trail {
private:
  TermVar *tcar;
  Trail *tcdr;
  static Trail *sofar;
  Trail(TermVar *h, Trail *t) : tcar(h), tcdr(t) {}

public:
  static Trail *Note() { return sofar; }
  static void Push(TermVar *x) { sofar = new Trail(x, sofar); }
  static void Undo(Trail *whereto) {
    for (; sofar != whereto; sofar = sofar->tcdr)
      sofar->tcar->reset();
  }
};
Trail *Trail::sofar = NULL;

bool TermVar::unify(Term *t) {
  if (instance != this)
    return instance->unify(t);
  Trail::Push(this);
  instance = t;
  return true;
}
Term *TermVar::copy() {
  if (instance == this) {
    Trail::Push(this);
    instance = new TermVar();
  }
  return instance;
}

class TermVarMapping {
private:
  TermVar **varvar;
  char **vartext;
  int size;

public:
  TermVarMapping(TermVar *vv[], char *vt[], int vs)
      : varvar(vv), vartext(vt), size(vs) {}
  void showanswer() {
    if (size == 0)
      cout << "yes\n";
    else {
      for (int i = 0; i < size; i++) {
        cout << vartext[i] << " = ";
        varvar[i]->print();
        cout << "\n";
      }
    }
  }
};

void Goal::solve(Program *p, int level, TermVarMapping *map) {
  indent(level .4);
  cout << "solve@" << level << ": ";
  this->print();
  cout << "\n";
  for (Program *q = p; q != NULL; q = q->pcdr) {
    Trail *t = Trail::Note();
    Clause *c = q->pcar->copy();
    Trail::Undo(t);
    indent(level, 4);
    cout << "  try:";
    c->print();
    cout << "\n";
    if (car->unify(c->head)) {
      Goal *gdash = c->body == NULL ? cdr : c->body->append(cdr);
      if (gdash == NULL)
        map->showanswer();
      else
        gdash->solve(p, level + 1, map);
    } else {
      indent(level, 4);
      cout << "  nomatch.\n";
    }
    Trail::Undo(t);
  }
}

/* A sample test program: append */

Atom *at_app = new Atom("app");
Atom *at_cons = new Atom("cons");
TermConstant *f_nil = new TermCons(new Atom("nil"));
TermConstant *f_1 = new TermCons(new Atom("1"));
TermConstant *f_2 = new TermCons(new Atom("2"));
TermConstant *f_3 = new TermCons(new Atom("3"));

Term *v_x = new TermVar();
TermConstant *lhs1 = new TermCons(at_app, f_nil, v_x, v_x);
Clause *c1 = new Clause(lhs1, NULL);

Term *v_l = new TermVar();
Term *v_m = new TermVar();
Term *v_n = new TermVar();
TermConstant *rhs2 = new TermCons(at_app, v_l, v_m, v_n);
TermConstant *lhs2 = new TermCons(at_app, new TermCons(at_cons, v_x, v_l), v_m,
                                  new TermCons(at_cons, v_x, v_n));
Clause *c2 = new Clause(lhs2, new Goal(rhs2, NULL));

TermVariable *v_i = new TermVar();
TermVariable *v_j = new TermVar();
TermConstant *rhs3 =
    new TermCons(at_app, v_i, v_j,
                 new TermCons(at_cons, f_1,
                              new TermCons(at_cons, f_2,
                                           new TermCons(at_cons, f_3, f_nil))));

Goal *g1 = new Goal(rhs3, NULL);

Program *test_p = new Program(c1, new Program(c2, NULL));
Program *test_p2 = new Program(c2, new Program(c1, NULL));

TermVariable *varvar[] = {v_i, v_j};
char *varname[] = {"I", "J"};
TermVarMapping *var_name_map = new TermVarMapping(varvar, varname, 2);

int main(int argc, char *argv[]) {
  cout << "=======Append with normal clause order:\n";
  g1->solve(test_p, 0, var_name_map);
  cout << "\n=======Append with reversed normal clause order:\n";
  g1->solve(test_p2, 0, var_name_map);
  return 0;
}

File: prolog.py
#!/usr/bin/env python3

class TermConstant:
    def __init__(self, fsym, *args):
        self.fsym = fsym
        self.args = list(args)

    def __repr__(self):
        rep = '('
        for i in range(0, len(self.args)):
            rep += (", " if i else "") + self.args[i]
        rep += ')'
        return rep

    def unify(self, term):
        return term.unifyCons(self)

    def unifyCons(self, cons):
        if not ((self.fsym == cons.fsym) and (len(self.args) == len(cons.args))):
            return False
        for i in range(self.args):
            if not args.unify(cons.args[i]):
                return false
        return True

    def copy(self):
        pass


_tvtimestamp = 0


class TermVar:
    def __init__(self, instance=0):
        global _tvtimestamp
        self.instance = self if not instance else instance
        self.timestamp = _tvtimestamp = _tvtimestamp + 1

    def __repr__(self):
        return repr(self.instance) if self.instance == self else "TermVar(" + self.timestamp + ")"

    def unify(self, term):
        if (self.instance != self):
            return instance.unify(term)
        instance = term
        # Push self to trail
        trail_push(self)
        self.instance = term
        return True

    def unifyCons(self, cons):
        return self.unify(cons)

    def copy(self):
        trail_push(self)
        return TermVar(self.instance)


class Goal:
    def __init__(self, car, cdr):
        self.car = car
        self.cdr = cdr
        pass

    def __repr__(self):
        return repr(self.car) + "" if not cdr else ('; ' + repr(cdr))

    def copy(self):
        return Goal(self.car, 0 if self.cdr else cdr.copy())

    def solve(self, prog, level, tvmap):
        pass


class Clause:
    def __init__(self, head, body):
        self.head = head
        self.body = body

    def __repr__(self):
        return repr(self.head) + " :- " + (repr(self.body) if self.body else 'true')


class TermVarMapping:
    pass


class Trail:
    def __init__(self, head, body):
        self.tcar = head
        self.tcdr = body
        pass

trail = Trail()

def trail_push(tv):
    global trail
    trail = Trail(tv, trail)

def trail_undo(to):
    global trail
    while trail != to:
        trail.tcar.instance = trail.tcar
        trail = trail.tcdr


class Program:
    def __init__(self, pcar, pcdr):
        self.pcar = pcar
        self.pcdr = pcdr


at_app = "app"
at_cons = "cons"
f_nil = TermConstant("nil")
f_1 = TermConstant("1")
f_2 = TermConstant("2")
f_3 = TermConstant("3")

v_x = TermVar()
lhs1 = TermConstant(at_app, f_nil, v_x, v_x)
c1 = Clause(lhs1, 0)

v_l = TermVar()
v_m = TermVar()
v_n = TermVar()
rhs2 = TermConstant(at_app, v_l, v_m, v_n)
lhs2 = TermConstant(at_app, TermConstant(at_cons,
                                 v_x, v_l), v_m,
                TermConstant(at_cons, v_x, v_n))
c2 = Clause(lhs2,  Goal(rhs2, 0))

v_i = TermVar()
v_j = TermVar()
rhs3 = TermConstant(at_app, v_i, v_j,
                TermConstant(at_cons, f_1,
                         TermConstant(at_cons, f_2,
                                  TermConstant(at_cons, f_3, f_nil))))

g1 = Goal(rhs3, 0)

test_p = Program(c1,  Program(c2, 0))
test_p2 = Program(c2,  Program(c1, 0))

varvar = [v_i, v_j]
varname = ["I", "J"]
var_name_map = TermVarMapping(varvar, varname, 2)
